File: agent.js
#!/usr/bin/env node
const { spawnSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const args = process.argv.slice(2);
const command = args[0];
const repo = args[1];
const update = args.includes('--update') || args.includes('-u');

function usage() {
  console.log('Usage: agent add <repo-url> [--update]');
  process.exit(1);
}

if (command !== 'add' || !repo) {
  usage();
}

const pluginsDir = path.resolve(__dirname, 'plugins');
if (!fs.existsSync(pluginsDir)) {
  fs.mkdirSync(pluginsDir, { recursive: true });
}

const repoName = path.basename(repo, '.git').split('/').pop();
const pluginPath = path.join(pluginsDir, repoName);

if (fs.existsSync(pluginPath)) {
  if (update) {
    console.log(`Updating existing plugin ${repoName}...`);
    const pull = spawnSync('git', ['pull'], { cwd: pluginPath, stdio: 'inherit' });
    if (pull.status !== 0) {
      console.error(`Failed to update plugin ${repoName}.`);
      process.exit(pull.status || 1);
    }
  } else {
    console.log(`Plugin ${repoName} already exists. Use --update to pull latest changes.`);
    process.exit(0);
  }
} else {
  const clone = spawnSync('git', ['clone', repo], { cwd: pluginsDir, stdio: 'inherit' });
  if (clone.status !== 0) {
    console.error(`Failed to clone repository ${repo}.`);
    process.exit(clone.status || 1);
  }
}

// copy frontend part if exists
const pluginSrc = path.join(pluginPath, 'src');
const targetSrc = path.resolve(__dirname, 'src/plugins', repoName);
if (fs.existsSync(pluginSrc)) {
  try {
    fs.cpSync(pluginSrc, targetSrc, { recursive: true });
  } catch (err) {
    console.error(`Failed to copy frontend files for ${repoName}:`, err.message);
  }
}

// reload ts registry
require('ts-node/register');
const { reloadRegistry } = require('./src/tools/registry');
reloadRegistry();

// reload python registry
spawnSync('python', ['-c', 'from py.tool_registry import reload_plugins; reload_plugins()'], { stdio: 'inherit' });
const pyReload = spawnSync('python', ['-c', 'from py.agent_tool import reload_registry; reload_registry()'], { stdio: 'inherit' });
if (pyReload.status !== 0) {
  console.error('Failed to reload python registry.');
  process.exit(pyReload.status || 1);
}

File: agent.py
import argparse
import subprocess
from pathlib import Path
from py.tool_registry import load_plugins

def add(repo_url: str, plugins_dir: str = "plugins") -> None:
    """Clone a plugin repository and activate its tools."""
    plugins_path = Path(plugins_dir)
    plugins_path.mkdir(exist_ok=True)
    dest = plugins_path / Path(repo_url).name
    subprocess.run([
        "git",
        "clone",
        repo_url,
        str(dest)
    ], check=True)
    load_plugins(str(plugins_path))

def main() -> None:
    parser = argparse.ArgumentParser(prog="agent")
    sub = parser.add_subparsers(dest="command")
    add_p = sub.add_parser("add")
    add_p.add_argument("repo_url")
    args = parser.parse_args()
    if args.command == "add":
        add(args.repo_url)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
